org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java

```diff
@@ -235,7 +235,7 @@ public static class Builder<T> {
 		protected boolean validateMessages;
 		protected Consumer<GsonBuilder> configureGson;
 		protected ClassLoader classLoader;
-		protected MessageTracer messageTracer;
+		protected PrintWriter messageTracer;
```

Comment:
This is a public class and thereby a breaking change. Is there a way to retain the MessageTracer type for this field?

Reply:
Unfortunately, this is not the only breaking API change here.

The culprit seems to be the `wrapMessageConsumer` method, which now requires `MessageJsonHandler` as the second parameter.

We could, of course, deprecate the old one-parameter method, but since it would no longer be called, there doesn't seem to be much value in it.

We could also deprecate the `messageTracer` field, but similarly, since it would no longer be used, would there be much value in it?

Putting the breaking change aside, the `PrintWriter` looks like a better type for the `messageTracer` to me. It follows the pattern established by other similar fields in the `Launcher.Builder`, where the type of the field matches the parameter type of the corresponding setter.

Having said that, it would be best to find a clean non-breaking solution to this problem, of course. Any suggestions would be very welcome.

Reply:
OK. How about this:

```java
	public static class Builder<T> {
		...
		protected PrintWriter tracer;
		protected MessageTracer messageTracer;
		...		
		public Builder<T> traceMessages(PrintWriter tracer) {
			if (tracer != null) {
				this.tracer = tracer;
				messageTracer = new MessageTracer(tracer); // fallback to old behavior
			}
			return this;
		}
		...
		public Launcher<T> create() {
			...
			MessageJsonHandler jsonHandler = createJsonHandler();
			messageTracer = new MessageTracer(tracer, jsonHandler);
			...
		}
		...
		// same as before -- no need to change anything in this method
		protected MessageConsumer wrapMessageConsumer(MessageConsumer consumer) {
			MessageConsumer result = consumer;
			if (messageTracer != null) {
				result = messageTracer.apply(consumer);
			}
			...
		}
		...
	}
```

Note that if a subclass completely overrides (i.e. replaces) the `Builder.create()` method, like `WebSocketLauncherBuilder` does in LSP4J, it will need to be modified to set the 'enhanced' `messageTracer`. Otherwise, it will fallback to the old behaviour.

Reply:
I made an attempt to not break any public API. The downside is that the jsonHandler will be set on the MessageTracer later on. Not sure if this could have any bad consequence.